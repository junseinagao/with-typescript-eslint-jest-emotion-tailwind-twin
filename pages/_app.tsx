import { GlobalStyles } from 'twin.macro'
import { AppProps } from 'next/app'

const App = ({ Component, pageProps }: AppProps) => (
  <div>
    <GlobalStyles />
    <Component {...pageProps} />
  </div>
)

export default App
