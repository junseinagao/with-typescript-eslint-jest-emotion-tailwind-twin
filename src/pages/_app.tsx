import Head from 'next/head'
import { GlobalStyles, css } from 'twin.macro'
import { Global } from '@emotion/react'
import { AppProps } from 'next/app'

const App = ({ Component, pageProps }: AppProps) => (
  <>
    <Head>
      <title>
        Nextjs App with TypeScript, ESlint, Jest, Emotion, Tailwind and Twin
      </title>
      <link rel="icon" href="/favicon.ico" />
    </Head>
    <GlobalStyles />
    <Global styles={globalStyles} />
    <Component {...pageProps} />
  </>
)

const globalStyles = css`
  html,
  body {
    padding: 0;
    margin: 0;
    font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
      Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
  }

  * {
    box-sizing: border-box;
  }
`

export default App
