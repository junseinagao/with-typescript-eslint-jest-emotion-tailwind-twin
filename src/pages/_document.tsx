import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
} from 'next/document'
import { extractCritical } from '@emotion/server'

export default class MyDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const initialProps = await Document.getInitialProps(ctx)
    const page = await ctx.renderPage()
    const styles = extractCritical(page.html)
    return {
      ...initialProps,
      ...page,
      styles: (
        <>
          {initialProps.styles}
          <style
            data-emotion-css={styles.ids.join(' ')}
            dangerouslySetInnerHTML={{ __html: styles.css }}
          />
        </>
      ),
    }
  }

  render() {
    return (
      <Html lang="en">
        <Head>
          <title>
            Nextjs App with TypeScript, ESlint, Jest, Emotion, Tailwind and Twin
          </title>
          <link rel="icon" href="/favicon.ico" />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}
