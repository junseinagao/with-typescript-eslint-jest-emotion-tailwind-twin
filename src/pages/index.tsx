import Image from 'next/image'
import tw, { css } from 'twin.macro'

export const Index = () => (
  <div css={container}>
    <section css={main}>
      <h1 css={[title, title_and_description]}>
        Welcome to <a href="https://nextjs.org">Next.js!</a>
      </h1>

      <p css={[description, title_and_description]}>
        Get started by editing <code>pages/index.tsx</code>
      </p>

      <button
        css={testButton}
        onClick={() => {
          window.alert(
            'With TypeScript, ESlint, Jest, Emotion, Tailwind and Twin'
          )
        }}
      >
        Test Button
      </button>

      <div css={grid}>
        <a href="https://nextjs.org/docs" css={[a, card]}>
          <h3>Documentation &rarr;</h3>
          <p>Find in-depth information about Next.js features and API.</p>
        </a>

        <a href="https://nextjs.org/learn" css={[a, card]}>
          <h3>Learn &rarr;</h3>
          <p>Learn about Next.js in an interactive course with quizzes!</p>
        </a>

        <a
          href="https://github.com/vercel/next.js/tree/master/examples"
          css={[a, card]}
        >
          <h3>Examples &rarr;</h3>
          <p>Discover and deploy boilerplate example Next.js projects.</p>
        </a>

        <a
          href="https://vercel.com/new?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
          css={[a, card]}
        >
          <h3>Deploy &rarr;</h3>
          <p>Instantly deploy your Next.js site to a public URL with Vercel.</p>
        </a>
      </div>
    </section>

    <footer css={footer}>
      <a
        href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
        target="_blank"
        rel="noopener noreferrer"
        css={a}
      >
        <span tw="mr-2">Powered by</span>
        <Image src="/vercel.svg" alt="Vercel Logo" height={'32'} width={'64'} />
      </a>
    </footer>
  </div>
)

const container = css`
  min-height: 100vh;
  padding: 0 0.5rem;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

const main = css`
  padding: 5rem 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

const footer = css`
  width: 100%;
  height: 100px;
  border-top: 1px solid #eaeaea;
  display: flex;
  justify-content: center;
  align-items: center;

  & img {
    margin-left: 0.5rem;
  }

  & a {
    display: flex;
    justify-content: center;
    align-items: center;
  }
`

const testButton = css`
  ${tw`mt-8 bg-blue-500 text-white rounded px-2 py-1`}
  font-family: Menlo, Monaco, Lucida Console, Liberation Mono, DejaVu Sans Mono, Bitstream Vera Sans Mono, Courier New, monospace;
`

const a = css`
  color: inherit;
  text-decoration: none;
`

const title = css`
  margin: 0;
  line-height: 1.15;
  font-size: 4rem;

  & a {
    color: #0070f3;
    text-decoration: none;
  }

  & a:hover,
  & a:focus,
  & a:active {
    text-decoration: underline;
  }
`

const title_and_description = css`
  text-align: center;
`

const description = css`
  line-height: 1.5;
  font-size: 1.5rem;
`

const grid = css`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;

  max-width: 800px;
  margin-top: 3rem;

  @media (max-width: 600px) {
    width: 100%;
    flex-direction: column;
  }
`

const card = css`
  margin: 1rem;
  flex-basis: 45%;
  padding: 1.5rem;
  text-align: left;
  color: inherit;
  text-decoration: none;
  border: 1px solid #eaeaea;
  border-radius: 10px;
  transition: color 0.15s ease, border-color 0.15s ease;

  &:hover,
  &:focus,
  &:active {
    color: #0070f3;
    border-color: #0070f3;
  }

  & h3 {
    margin: 0 0 1rem 0;
    font-size: 1.5rem;
  }

  & p {
    margin: 0;
    font-size: 1.25rem;
    line-height: 1.5;
  }
`

export default Index
