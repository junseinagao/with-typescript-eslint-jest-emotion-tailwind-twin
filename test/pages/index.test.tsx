import React from 'react'
import { render, fireEvent } from '../testUtils'
import { Index } from '~/pages'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Index />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking button triggers alert', () => {
    const { getByText } = render(<Index />, {})
    window.alert = jest.fn()
    fireEvent.click(getByText('Test Button'))
    expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  })
})
