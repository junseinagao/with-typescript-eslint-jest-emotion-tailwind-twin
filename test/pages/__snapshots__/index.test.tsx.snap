// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-18dlair-container"
  >
    <section
      class="css-1kcqiws-main"
    >
      <h1
        class="css-x8g5az-title-title_and_description-Index"
      >
        Welcome to 
        <a
          href="https://nextjs.org"
        >
          Next.js!
        </a>
      </h1>
      <p
        class="css-ls6h2g-description-title_and_description-Index"
      >
        Get started by editing 
        <code>
          pages/index.tsx
        </code>
      </p>
      <button
        class="css-1qntr4c-testButton"
      >
        Test Button
      </button>
      <div
        class="css-1yqhy4u-grid"
      >
        <a
          class="css-1sek87q-a-card-Index"
          href="https://nextjs.org/docs"
        >
          <h3>
            Documentation →
          </h3>
          <p>
            Find in-depth information about Next.js features and API.
          </p>
        </a>
        <a
          class="css-1sek87q-a-card-Index"
          href="https://nextjs.org/learn"
        >
          <h3>
            Learn →
          </h3>
          <p>
            Learn about Next.js in an interactive course with quizzes!
          </p>
        </a>
        <a
          class="css-1sek87q-a-card-Index"
          href="https://github.com/vercel/next.js/tree/master/examples"
        >
          <h3>
            Examples →
          </h3>
          <p>
            Discover and deploy boilerplate example Next.js projects.
          </p>
        </a>
        <a
          class="css-1sek87q-a-card-Index"
          href="https://vercel.com/new?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
        >
          <h3>
            Deploy →
          </h3>
          <p>
            Instantly deploy your Next.js site to a public URL with Vercel.
          </p>
        </a>
      </div>
    </section>
    <footer
      class="css-14ixi6n-footer"
    >
      <a
        class="css-dszpow-a"
        href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="css-1oywshl-Index"
        >
          Powered by
        </span>
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iMzIiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Vercel Logo"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
